servlet/src/main/java/io/undertow/servlet/handlers/ServletPathMatchesData.java

```diff
@@ -79,22 +80,22 @@ public ServletPathMatch getServletHandlerByPath(final String path) {
             }
         }
         //this should never happen
-        //as the default servlet is aways registered under /*
+        //as the default servlet is always registered under /*
         throw UndertowMessages.MESSAGES.servletPathMatchFailed();
     }
 
     private ServletPathMatch handleMatch(final String path, final PathMatch match, final int extensionPos) {
-        if (match.extensionMatches.isEmpty()) {
+        if (extensionPos == -1 || match.extensionMatches.isEmpty()) {
             return new ServletPathMatch(match.defaultHandler, path, match.requireWelcomeFileMatch);
         }
-        if (extensionPos == -1) {
-            return new ServletPathMatch(match.defaultHandler, path, match.requireWelcomeFileMatch);
-        }
-        final String ext;
-        ext = path.substring(extensionPos + 1, path.length());
-        ServletChain handler = match.extensionMatches.get(ext);
-        if (handler != null) {
-            return new ServletPathMatch(handler, path, handler.getManagedServlet().getServletInfo().isRequireWelcomeFileMapping());
+        MappingMatch matchType = match.defaultHandler.getMappingMatch();
+        // Extension matches take precedence over the default servlet, but not path or exact matches
+        if (matchType != MappingMatch.PATH && matchType != MappingMatch.EXACT) {
```

Comment:
Here we wrap the existing code such that EXTENSION mapping doesn't occur when we have any PATH or EXACT match, based on the servlet specification.

Reply:
Ahh this fails due to filters, which can match both. I think I'm looking at the wrong layer. ServletPathMatches.java looks like the correct place to resolve this issue.